File: src/main/java/net/minestom/server/utils/collection/IntMappedArray.java
package net.minestom.server.utils.collection;

import org.jetbrains.annotations.ApiStatus;

import java.util.AbstractList;
import java.util.Objects;
import java.util.function.IntFunction;

@ApiStatus.Internal
public final class IntMappedArray<R> extends AbstractList<R> {
    private final int[] elements;
    private final IntFunction<R> function;

    public IntMappedArray(int[] elements, IntFunction<R> function) {
        this.elements = elements;
        this.function = function;
    }

    @Override
    public R get(int index) {
        final int[] elements = this.elements;
        Objects.checkIndex(index, elements.length);
        return function.apply(elements[index]);
    }

    @Override
    public int size() {
        return elements.length;
    }
}

File: src/main/java/net/minestom/server/utils/collection/IntMappedArray.kt
package net.minestom.server.utils.collection

import java.util.AbstractMap.SimpleImmutableEntry
import java.util.function.BiConsumer
import java.util.function.IntFunction
import java.lang.IndexOutOfBoundsException
import java.util.concurrent.atomic.AtomicReference
import net.minestom.server.utils.collection.MappedCollection
import org.jetbrains.annotations.ApiStatus
import java.util.AbstractList

@ApiStatus.Internal
class IntMappedArray<R>(private val elements: IntArray, private val function: IntFunction<R>) : AbstractList<R>() {
    override fun get(index: Int): R {
        val elements = elements
        if (index < 0 || index >= elements.size) throw IndexOutOfBoundsException("Index " + index + " is out of bounds for length " + elements.size)
        return function.apply(elements[index])
    }

    override fun size(): Int {
        return elements.size
    }
}